arbor/backends/gpu/matrix_fine.cu

```diff
@@ -233,15 +233,17 @@ void solve_matrix_fine(
 
             // Find the index of the first node in this branch.
             const unsigned len = lvl_lengths[tid];
-            unsigned pos = lvl_meta.matrix_data_index + (len-1)*width + tid;
+            const int pos0 = lvl_data_index + tid;
+            const int posN = pos0 + (len - 1)*width;
```

Comment:
This new approach to indexing makes the for loop logic more complex. I prefer the old approach which explicitly loops over the number of branches. It is clearer and less likely to be buggy.

Reply:
While refactoring this I felt the exact opposite as it's hard to keep track of the current state of `pos`. This way it is always a straight counting loop and `pos` is not kept alive in between loops. That said, I wholeheartedly agree that the loop boundaries muddle the waters and could be better motivated.